File: 720.cpp
//
//  Solution.cpp
//  LeetCode
//


#include "Solution.hpp"
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <unordered_set>

class Solution {
public:
    std::string longestWord(std::vector<std::string>& words) {
        std::string res;
        std::unordered_set<std::string> s;
        
        sort(words.begin(), words.end());
        
        for (std::string t : words) {
            if (t.length() == 1 || s.count(t.substr(0, t.length() - 1)) > 0) {
                s.emplace(t);
                if (t.length() > res.length()) {
                    res = t;
                }
            }
        }
        
        return res;
    }
};

File: 720.java
import java.util.*;

class Solution {
    public String longestWord(String[] words) {
        Arrays.sort(words);
        Set<String> s = new HashSet<>();
        String res = new String();

        for (String t : words) {
            if (t.length() == 1 || s.contains(t.substring(0, t.length() - 1)) == true) {
                s.add(t);
                if (t.length() > res.length()) {
                    res = t;
                }
            }
        }

        return res;
    }
}
